src/attr/_funcs.py

```diff
@@ -337,8 +338,13 @@ def evolve(inst, **changes):
             continue
         attr_name = a.name  # To deal with private attributes.
         init_name = attr_name if attr_name[0] != "_" else attr_name[1:]
+        value = getattr(inst, attr_name)
         if init_name not in changes:
-            changes[init_name] = getattr(inst, attr_name)
+            # Add original value to changes
+            changes[init_name] = value
+        elif has(value):
+            # Evolve nested attrs classes
+            changes[init_name] = evolve(value, **changes[init_name])
```

Comment:
doesn't that break the edge case of legitimately replacing a attrs instance with a dict where  permissible?

Reply:
`has(value)` makes sure we only attempt recursion when the current value is an `attrs` class.

I added a test asserting that dicts can still normaly be replaced, nonetheless. :)